### PR TITLE
[chip,dv] flash_ctrl memory protection test

### PR DESCRIPTION
**Pull request by jdonjdon**
Test for flash_ctrl multi mp regions and their priority.
Test verified in simulation platform.
sival target will be added in the subsequent PR.

### CODE REVIEW

File: sw/device/tests/flash_ctrl_mem_protection_test.c
Comment (matutem) [inline sw/device/tests/flash_ctrl_mem_protection_test.c:22]:
```suggestion
 * any regions overlap.
```

File: sw/device/tests/flash_ctrl_mem_protection_test.c
Comment (matutem) [inline sw/device/tests/flash_ctrl_mem_protection_test.c:32]:
```suggestion
 * As you see, these regions intentionally overlap each other.
```

File: sw/device/tests/flash_ctrl_mem_protection_test.c
Comment (matutem) [inline sw/device/tests/flash_ctrl_mem_protection_test.c:33]:
```suggestion
 * For overlapping regions the lower region has priority.
```

File: sw/device/tests/flash_ctrl_mem_protection_test.c
Comment (matutem) [inline sw/device/tests/flash_ctrl_mem_protection_test.c:202]: I am curious: should the test have two loops here, one that sets all region properties, and another one that does test_mem_access? This way test_mem_access will be certain to run with properties set as described above.

Comment (jdonjdon) [reply sw/device/tests/flash_ctrl_mem_protection_test.c:202]: Each loop it programs [table entry](https://github.com/lowRISC/opentitan/blob/2bfcf2a012698672160582861d5f1b38bb720c01/sw/device/tests/flash_ctrl_mem_protection_test.c#L26))
to cover multiple MP regions.

File: sw/device/tests/flash_ctrl_mem_protection_test.c
Comment (jdonjdon) [reply sw/device/tests/flash_ctrl_mem_protection_test.c]: fixed.

File: sw/device/tests/flash_ctrl_mem_protection_test.c
Comment (jdonjdon) [reply sw/device/tests/flash_ctrl_mem_protection_test.c]: fixed

**Review by hcallahan-lowrisc — COMMENTED** [review]
Test LGTM, though I would wait for another more-familiar reviewer to approve as well. 

One question. Is it worth covering bank 0 as well in this test?

File: sw/device/tests/flash_ctrl_mem_protection_test.c
Comment (matutem) [reply sw/device/tests/flash_ctrl_mem_protection_test.c:202]: I think I understand: each time you create a new region overlapping previous regions starting at a lower page, so it has higher priority. If you agree this is how it works, please change the text around line 34 to reflect this.

Comment (jdonjdon) [reply sw/device/tests/flash_ctrl_mem_protection_test.c:202]: @matutem , test program higher region first, which has "lower" priority.
Then subsequent programming with lower region will overruled previous config parameter because
lower region number has "higher" priority.

**Review by matutem — APPROVED** [review]
Jaedon, I added a comment with my understanding of the logic of the test. If it is accurate it can be used to enhance the comments near line 34 of the .c file, since the test is not obvious otherwise.

I'll leave it up to you to decide what to do.

LGTM

**Comment by jdonjdon** [comment]
> Jaedon, I added a comment with my understanding of the logic of the test. If it is accurate it can be used to enhance the comments near line 34 of the .c file, since the test is not obvious otherwise.
> 
> I'll leave it up to you to decide what to do.
> 
> LGTM

Thanks, @matutem Comment has been updated as suggested.

**Pull request merged**
